pulsar/schema.go

```diff
@@ -0,0 +1,502 @@
+	return as.Decode(message, nil)
+}
+
+func (as *AvroSchema) GetSchemaInfo() *SchemaInfo {
+	return &as.SchemaInfo
+}
+
+type StringSchema struct {
+	SchemaInfo
+}
+
+func NewStringSchema(properties map[string]string) *StringSchema {
+	strSchema := new(StringSchema)
+	strSchema.SchemaInfo.Properties = properties
+	strSchema.SchemaInfo.Name = "String"
+	strSchema.SchemaInfo.Type = STRING
+	strSchema.SchemaInfo.Schema = ""
+	return strSchema
+}
+
+func (ss *StringSchema) Encode(v interface{}) ([]byte, error) {
+	return []byte(v.(string)), nil
+}
+
+func (ss *StringSchema) Decode(data []byte, v interface{}) error {
+	bh := (*reflect.SliceHeader)(unsafe.Pointer(&data))
+	sh := reflect.StringHeader{
+		Data: bh.Data,
+		Len:  bh.Len,
+	}
+	shPtr := (*string)(unsafe.Pointer(&sh))
+	reflect.ValueOf(v).Elem().Set(reflect.ValueOf(shPtr))
+	return nil
+}
+
+func (ss *StringSchema) Validate(message []byte) error {
+	return ss.Decode(message, nil)
+}
+
+func (ss *StringSchema) GetSchemaInfo() *SchemaInfo {
+	return &ss.SchemaInfo
+}
+
+type BytesSchema struct {
+	SchemaInfo
+}
+
+func NewBytesSchema(properties map[string]string) *BytesSchema {
+	bytesSchema := new(BytesSchema)
+	bytesSchema.SchemaInfo.Properties = properties
+	bytesSchema.SchemaInfo.Name = "Bytes"
+	bytesSchema.SchemaInfo.Type = BYTES
+	bytesSchema.SchemaInfo.Schema = ""
+	return bytesSchema
+}
+
+func (bs *BytesSchema) Encode(data interface{}) ([]byte, error) {
+	return data.([]byte), nil
+}
+
+func (bs *BytesSchema) Decode(data []byte, v interface{}) error {
+	reflect.ValueOf(v).Elem().Set(reflect.ValueOf(data))
+	return nil
+}
+
+func (bs *BytesSchema) Validate(message []byte) error {
+	return bs.Decode(message, nil)
+}
+
+func (bs *BytesSchema) GetSchemaInfo() *SchemaInfo {
+	return &bs.SchemaInfo
+}
+
+type Int8Schema struct {
+	SchemaInfo
+}
+
+func NewInt8Schema(properties map[string]string) *Int8Schema {
+	int8Schema := new(Int8Schema)
+	int8Schema.SchemaInfo.Properties = properties
+	int8Schema.SchemaInfo.Schema = ""
+	int8Schema.SchemaInfo.Type = INT8
+	int8Schema.SchemaInfo.Name = "INT8"
+	return int8Schema
+}
+
+func (is8 *Int8Schema) Encode(value interface{}) ([]byte, error) {
+	var buf bytes.Buffer
+	err := WriteElements(&buf, value.(int8))
+	return buf.Bytes(), err
+}
+
+func (is8 *Int8Schema) Decode(data []byte, v interface{}) error {
+	buf := bytes.NewReader(data)
+	return ReadElements(buf, v)
+}
+
+func (is8 *Int8Schema) Validate(message []byte) error {
+	if len(message) != 1 {
+		return errors.New("size of data received by Int8Schema is not 1")
+	}
+	return nil
+}
+
+func (is8 *Int8Schema) GetSchemaInfo() *SchemaInfo {
+	return &is8.SchemaInfo
+}
+
+type Int16Schema struct {
+	SchemaInfo
+}
+
+func NewInt16Schema(properties map[string]string) *Int16Schema {
+	int16Schema := new(Int16Schema)
+	int16Schema.SchemaInfo.Properties = properties
+	int16Schema.SchemaInfo.Name = "INT16"
+	int16Schema.SchemaInfo.Type = INT16
+	int16Schema.SchemaInfo.Schema = ""
```

Comment:
Why is the schema "" for all primitives?

Reply:
For primitives types, the schema field should be `""`